Sources/NIOWebSocket/WebSocketFrameDecoder.swift

```diff
@@ -0,0 +1,321 @@
+    public init(maxFrameSize: Int = 1 << 14) {
+        precondition(maxFrameSize <= UInt32.max, "invalid overlarge max frame size")
+        self.maxFrameSize = maxFrameSize
+    }
+
+    public func decode(ctx: ChannelHandlerContext, buffer: inout ByteBuffer) -> DecodingState  {
+        // Even though the calling code will loop around calling us in `decode`, we can't quite
+        // rely on that: sometimes we have zero-length elements to parse, and the caller doesn't
+        // guarantee to call us with zero-length bytes.
+        parseLoop: while self.shouldKeepParsing {
+            switch parser.parseStep(&buffer) {
+            case .result(let frame):
+                ctx.fireChannelRead(self.wrapInboundOut(frame))
+            case .continueParsing:
+                do {
+                    try self.validateState()
+                } catch {
+                    self.handleError(error, ctx: ctx)
+                }
+            case .insufficientData:
+                break parseLoop
+            }
+        }
+
+        // We parse eagerly, so once we get here we definitionally need more data.
+        return .needMoreData
+    }
+
+    /// Apply a number of validations to the incremental state, ensuring that the frame we're
+    /// receiving is valid.
+    private func validateState() throws {
+        if let length = parser.length, length > self.maxFrameSize {
+            throw NIOWebSocketError.invalidFrameLength
+        }
+
+        if let length = parser.length, let firstByte = parser.firstByte {
+            let isControlFrame = (firstByte & 0x08) != 0
+            let isFragment = (firstByte & 0x80) == 0
+
+            if isControlFrame && isFragment {
+                throw NIOWebSocketError.fragmentedControlFrame
+            }
+            if isControlFrame && length > 125 {
+                throw NIOWebSocketError.multiByteControlFrameLength
+            }
+        }
+    }
+
+    /// We hit a decoding error, we're going to tear things down now. To do this we're
+    /// basically going to send an error frame and then close the connection. Once we're
+    /// in this state we do no further parsing.
+    ///
+    /// A clean websocket shutdown is not really supposed to have an immediate close,
+    /// but we're doing that because the remote peer has prevented us from doing
+    /// further frame parsing, so we can't really wait for the next frame.
+    private func handleError(_ error: Error, ctx: ChannelHandlerContext) {
+        guard let error = error as? NIOWebSocketError else {
+            fatalError("Can only handle NIOWebSocketErrors")
+        }
+        self.shouldKeepParsing = false
+
+        var data = ctx.channel.allocator.buffer(capacity: 2)
+        WebSocketErrorCode(error).write(to: &data)
+        let frame = WebSocketFrame(fin: true,
+                                   opcode: .connectionClose,
+                                   data: data)
+        _ = ctx.write(self.wrapOutboundOut(frame)).then {
```

Comment:
don't you want to `close` on error too?

Reply:
I'm making the assumption that if the write errors, we have already closed.